### ISSUE TITLE
Builder that implements interface with parameterized setter fails to satisfy setter requirements

### ISSUE THREAD

**Issue by andrewmunn** [google/auto#380]
Suppose you have the following AutoValue class:

``` java
@AutoValue
public abstract class Child implements Parent {

    public static Builder builder() {
        return new AutoValue_Child.Builder();
    }

    @Override
    public abstract List<Child> data();

    @AutoValue.Builder
    public abstract static class Builder implements Parent.Builder<Builder, Child> {

        public abstract Child build();
    }
}
```

that implements this interface:

``` java
public interface Parent  {

    List<? extends Parent> data();

    interface Builder<BUILDER extends Builder, TYPE extends Parent> {

        BUILDER data(List<TYPE> data);
    }
}
```

This fails to compile with:

`> Error:(11, 17) error: Parameter type of setter method should be java.util.List<com.example.testapp.Child> to match getter com.example.testapp.Child.data`

This seems like a bug because the setter's parameter is `TYPE` which in this resolves to `Child`, so everything _should_ compile just fine. In order to get AutoValue to compile, I must provide the setter in `Child`'s Builder like so:

``` java
 @AutoValue.Builder
    public abstract static class Builder implements Parent.Builder<Builder, Child> {
        public abstract Builder data(List<Child> data);
        public abstract Child build();
    }

```

**Comment by eamonnmcmanus** [comment]
Yes, this is a bug. We should use [Types.asMemberOf](https://docs.oracle.com/javase/8/docs/api/javax/lang/model/util/Types.html#asMemberOf-javax.lang.model.type.DeclaredType-javax.lang.model.element.Element-) here. Unfortunately, that method has historically not worked correctly on Eclipse (which I see you are using): [this bug](https://bugs.eclipse.org/bugs/show_bug.cgi?id=382590) has been fixed but [this one](https://bugs.eclipse.org/bugs/show_bug.cgi?id=499026) is still open. We have [a workaround](https://github.com/google/auto/blob/master/value/src/main/java/com/google/auto/value/processor/EclipseHack.java#L93) when the method has no parameters but not for the general case.

I think the best we can do for now is to fix the bug you describe in the non-Eclipse case, but in a way that at least does not blow up with an exception in Eclipse.

**Comment by andrewmunn** [comment]
@eamonnmcmanus actually this was in Android Studio.

**Comment by pettermahlen** [comment]
I wanted to use a property like

```
public abstract List<ChildType> children();
```

With a more permissive setter method in the builder:

```
public abstract Builder children(Iterable<? extends ChildType> newValue);
```

But I get the same error message. Is that the same issue, or am I perhaps doing something wrong?

A very interesting thing happened as I was writing this report: I changed the property to `ImmutableList<ChildType>`, and then suddenly it compiles just fine. Looks a little odd to me.

**Comment by pettermahlen** [comment]
@eamonnmcmanus makes perfect sense, thanks for the explanation!

**Comment by liujoshua** [comment]
I'm seeing this in Android Studio and command line gradle build. Is this the same issue?

Parameter type E of setter method should be E to match getter 

```
public interface DefaultAnswer<E {
    @Nullable
    E getDefaultAnswer();
}
public abstract class InputField<E> implements DefaultAnswer<E>{
    Builder<E> {
        abstract Builder<E> setDefaultAnswer(@Nullable final E defaultAnswer);
    }
}
```

I fixed the build by overriding getDefaultAnswer in InputField

```
public abstract class InputField<E> implements DefaultAnswer<E>{
    Builder<E> {
        abstract Builder<E> setDefaultAnswer(@Nullable final E defaultAnswer);
    }
    @Nullable
    @Override
    E getDefaultAnswer();
}
```

**Comment by eamonnmcmanus** [comment]
This issue should be fixed in 1.6.4, which has just been released.
